wsrep_connection_monitor_service.h

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (C) 2024-2925 Codership Oy <info@codership.com>
+ *
+ * This file is part of wsrep-API.
+ *
+ * Wsrep-API is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Wsrep-API is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with wsrep-API.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/** @file wsrep_connection_monitor_service.h
+ *
+ * This file defines interface for connection monitor service
+ *
+ * The provider which is capable of using the service interface v1 must
+ * export the following functions.
+ *
+ * int wsrep_init_connection_monitor_service_v1(wsrep_connection_monitor_service_v1_t*)
+ * void wsrep_deinit_connection_monitor_service_v1()
+ *
+ * which can be probed by the application.
+ *
+ * The application must initialize the service via above init function
+ * before the provider is initialized via wsrep->init(). The deinit
+ * function must be called after the provider side resources have been
+ * released via wsrep->free().
+ */
+
+#ifndef WSREP_CONNECTION_MONITOR_SERVICE_H
+#define WSREP_CONNECTION_MONITOR_SERVICE_H
+
+#include "wsrep_api.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif /* __cplusplus */
+
+/**
+ * Type tag for application defined connection monitoring processing context.
+ *
+ * Application may pass pointer to the context when initializing
+ * the connection monitor service. This pointer is passed a first parameter for
+ * each service call.
+ */
+typedef struct wsrep_connection_monitor_context wsrep_connection_monitor_context_t;
+
+typedef const void* wsrep_connection_key_t;
+
+/*
+ * Connection monitor connection update callbacks.
+ *
+ */
+
+/*
+ * Connection connect callback.
+ */
+typedef void (*wsrep_connection_monitor_connect_cb_t)(
+    wsrep_connection_monitor_context_t*,
+    wsrep_connection_key_t id,
+    const wsrep_buf_t* scheme,
+    const wsrep_buf_t* local_address,
```

Comment:
I would consider not to use single string to represent the addresses. E.g. with IPv6 addresses Galera uses the convention to enclose the address in square brackets to clearly separate the port, like in https://en.wikipedia.org/wiki/IPv6_address#Literal_IPv6_addresses_in_network_resource_identifiers.

It might be better to create a separate struct to represent address, like

```
typedef struct wsrep_address_t {
  wsrep_buf_t host;
  unsigned short port;
};
```
so that the host and port part can be easily looked up without parsing the address string.

Reply:
I have considered it, however I took here same approach as in allowlist i.e. addresses are single strings to make it possible to compare wsrep_allowlist contents to current connections.